internal/client/consensus/grandpa/communication_gossip.go

```diff
@@ -0,0 +1,1708 @@
+		}
+
+		broadcastTopics = topics
+		peerReply = catchUp
+		a = act
+	case gossipMessageCatchUp[H, N]:
+		messageName = "catch_up"
+		gv.innerMtx.Lock()
+		defer gv.innerMtx.Unlock()
+		a = gv.inner.validateCatchUpMessage(who, fullCatchUpMessage[H, N](val))
+	case gossipMessageCatchUpRequest:
+		messageName = "catch_up_request"
+		gv.innerMtx.Lock()
+		defer gv.innerMtx.Unlock()
+		reply, act := gv.inner.handleCatchUpRequest(who, catchUpRequestMessage(val), gv.setState)
+		peerReply = reply
+		a = act
+	default:
+		panic("unreachable")
+	}
+
+	// TODO: metrics
+	_ = messageName
+
+	return a, broadcastTopics, peerReply
+}
+
+func (gv *gossipValidator[H, N, Hasher]) NewPeer(
+	context gossip.ValidatorContext[H], who peerid.PeerID, roles role.ObservedRole,
+) {
+	gv.innerMtx.Lock()
+	gv.inner.peers.newPeer(who, roles)
+	gv.innerMtx.Unlock()
+
+	var packet *neighborPacket[N]
+	if gv.inner.localView != nil {
+		var height N
+		if h := gv.inner.localView.lastCommitHeight(); h != nil {
+			height = *h
+		}
+		packet = &neighborPacket[N]{
+			Round:                 gv.inner.localView.round,
+			SetID:                 gv.inner.localView.setID,
+			CommitFinalizedHeight: height,
+		}
+	}
+
+	if packet != nil {
+		var gossipMessage gossipMessageVDT[H, N]
+		setGossipMessageVDT(&gossipMessage, gossipMessageNeighbor[N](
+			versionedNeighborPacket[N]{neighborPacket: *packet}))
+
+		packetData := scale.MustMarshal(gossipMessage)
+		context.SendMessage(who, packetData)
+	}
+}
+
+func (gv *gossipValidator[H, N, Hasher]) PeerDisconnected(_ gossip.ValidatorContext[H], who peerid.PeerID) {
+	gv.innerMtx.Lock()
+	defer gv.innerMtx.Unlock()
+	gv.inner.peers.peerDisconnected(who)
+}
+
+func (gv *gossipValidator[H, N, Hasher]) Validate(
+	context gossip.ValidatorContext[H],
+	who peerid.PeerID,
+	data []byte,
+) gossip.ValidationResult {
+	action, broadcastTopics, peerReply := gv.doValidate(who, data)
+
+	// not with lock held!
+	if peerReply != nil {
+		var gossipMessage gossipMessageVDT[H, N]
+		gossipMessage.inner = peerReply
+		context.SendMessage(who, scale.MustMarshal(gossipMessage))
+	}
+
+	for _, topic := range broadcastTopics {
+		context.SendTopic(who, topic, false)
+	}
+
+	switch action := action.(type) {
+	case actionKeep[H]:
+		gv.report(who, action.ReputationChange)
+		context.BroadcastMessage(action.Hash, data, false)
+		return gossip.ValidationResultProcessAndKeep[H]{Hash: action.Hash}
+	case actionProcessAndDiscard[H]:
+		gv.report(who, action.ReputationChange)
+		return gossip.ValidationResultProcessAndDiscard[H]{Hash: action.Hash}
+	case actionDiscard[H]:
+		gv.report(who, action.ReputationChange)
+		return gossip.ValidationResultDiscard{}
+	default:
+		panic("unreachable")
+	}
+}
+
+func (gv *gossipValidator[H, N, Hasher]) MessageAllowed() func(
+	who peerid.PeerID,
+	intent gossip.MessageIntent,
+	topic H,
+	data []byte,
+) bool {
+	var doRebroadcast bool
+	{
+		gv.innerMtx.Lock()
+		defer gv.innerMtx.Unlock()
```

Comment:
This defer is not called in the end of the `{... }` block, given this example https://go.dev/play/p/DPPcivc3Zju the defer still being called at the function return/end

Reply:
Yea that was my understanding as well.  I don't think we even need the closure either, I'll remove it.